Add support for generators

jordyvanderhaegen opened the pull request:
Hi,

When retrieving large datasets from Exact this package will load all models into a single array, often resulting in memory overflow issues. This PR adds support for [generators](https://www.php.net/manual/en/class.generator.php) which avoids this issue by not loading the entire dataset in memory at once.

This has been PR'd by other people multiple times but has mostly been closed due to lack of tests. See #388 and #502 

**What has been changed in this PR?**

- The `Findable` trait now has a new `getGenerator` method
- The `get` method in the `Findable` trait calls this `getGenerator` method and returns it as an array. This is backwards compatible.
- The `Resultset` class now has a new `nextGenerator` method
- The `next` method in the `Resultset` class calls this `nextGenerator` method and returns it as an array. This is backwards compatible.
- The `collectionFromResult` method in both the `Findable` trait and `Resultset` class now return a Generator instead of an array. This looks to be a breaking change since this method is publicly available in `Findable`.

**What tests have been changed in this PR?**

- The `createMockConnection` and `createMockHandler` found in the `QueryParamTest` class have been extracted into a `MocksExactConnection` trait.
- This trait adds support for using fixtures, which should allow further contributors to include tests.
- The `get` and `getGenerator` methods of the `Findable` trait are now tested
- The `next` and `nextGenerator` methods of the `Resultset` class are now tested
- `.phpunit.result.cache` has been added to the gitignore

Hope this can be merged although it includes a breaking change 🤞

DannyvdSluijs commented on src/Picqer/Financials/Exact/Query/Findable.php:
```diff
@@ -169,13 +179,9 @@ public function collectionFromResult($result, array $headers = [])
-        foreach ($result as $r) {
-            $collection[] = new static($this->connection(), $r);
+            foreach ($nextResult as $row) {
+                yield new self($this->connection(), $row);
```
```suggestion
                yield new static($this->connection(), $row);
```
I guess we should keep using the late static binding method to avoid issues in the inheritance tree.

DannyvdSluijs commented on tests/ModelTest.php:
```diff
@@ -0,0 +1,36 @@
+        $this->assertCount(2, $response);
+    }
+
+    /** @test */
```
```suggestion
```
This annotation seems redundant as the test name starts with `test`

DannyvdSluijs reviewed (approved):
Cool to see a complete generator PR 👍🏼 
I gave it a review (didn't do any functional testing) but looks good to me.
Left some small comments for you to decide upon.

jordyvanderhaegen commented:
I'll give this a shot in production to see if everything still works as expected

jordyvanderhaegen commented:
I've changed some stuff in my latest commit:

- Renamed `xxxGenerator` methods to `xxxAsGenerator`
- Added a separate generator method for `filter` and `collectionFromResult`. Which should make this PR not have any breaking changes 🎉
- Added tests for all of the above + the `find` and `first` methods
